bootstore/tests/v0-fsm-proptest-learner.rs

```diff
@@ -0,0 +1,479 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Property based test for bootstore scheme v0 protocol logic
+//!
+//! This test creates a single `Fsm` as the system under test (SUT), where that
+//! FSM is a learner node.
+
+mod common;
+
+use assert_matches::assert_matches;
+use bootstore::schemes::v0::{
+    create_pkgs, ApiError, ApiOutput, Config, Envelope, Fsm, LearnedSharePkg,
+    Msg, MsgError, RackUuid, Request, RequestType, Response, ResponseType,
+    Share, SharePkg, State,
+};
+use bootstore::trust_quorum::RackSecret;
+use common::CommonTestState;
+use proptest::prelude::*;
+use secrecy::ExposeSecret;
+use sled_hardware::Baseboard;
+use std::collections::{BTreeMap, BTreeSet};
+use uuid::Uuid;
+
+use common::generators::{
+    arb_action, arb_config, arb_initial_member_ids, arb_learner_id, Action,
+    MAX_ACTIONS, MAX_INITIAL_MEMBERS, MIN_INITIAL_MEMBERS, TICK_TIMEOUT,
+};
+
+/// Actions run during the learning phase of the test
+#[derive(Debug, Clone)]
+pub enum LearnAction {
+    Timeout(Baseboard),
+    CannotSpareAShare(Baseboard),
+    Success(Baseboard),
+}
+
+#[derive(Debug, Clone)]
+pub struct TestInput {
+    pub sut_id: Baseboard,
+    pub initial_members: BTreeSet<Baseboard>,
+    pub config: Config,
+    pub rack_uuid: RackUuid,
+    pub learn_sequence: Vec<LearnAction>,
+    pub actions: Vec<Action>,
+}
+
+/// We randomize the order of learners, and only allow one node to distribute
+/// shares, thus ensuring that the learner itself will rotate and exercise the
+/// full learning path.
+fn arb_learn_sequence(
+    initial_members: BTreeSet<Baseboard>,
+) -> impl Strategy<Value = Vec<LearnAction>> {
+    // The peer that successfully responds to the learner
+    let peers: Vec<_> = initial_members.iter().cloned().collect();
+    (
+        any::<prop::sample::Index>(),
+        any::<prop::sample::Index>(),
+        Just(peers).prop_shuffle(),
+    )
+        .prop_map(|(success_index, timeout_index, peers)| {
+            let success_index = success_index.index(peers.len());
+            let timeout_index = timeout_index.index(peers.len());
+            let mut seq = vec![];
+            for (i, peer) in peers.into_iter().enumerate() {
+                if i == success_index {
+                    seq.push(LearnAction::Success(peer));
+                    break;
+                } else if i >= timeout_index {
+                    seq.push(LearnAction::Timeout(peer));
+                } else {
+                    seq.push(LearnAction::CannotSpareAShare(peer));
+                }
+            }
+            seq
+        })
+}
+
+/// Create the input to this test
+fn arb_test_input() -> impl Strategy<Value = TestInput> {
+    arb_initial_member_ids(MIN_INITIAL_MEMBERS, MAX_INITIAL_MEMBERS)
```

Comment:
This can be written without `prop_flat_map` if you're interested, it would involve using indexes everywhere and then resolving them afterwards. Likely not worth doing it though (esp after you add `no_shrink`).